annotator-core/src/main/java/edu/ucr/cs/riple/core/injectors/VirtualInjector.java

```diff
@@ -75,33 +82,57 @@ public void injectAnnotations(Set<AddAnnotation> changes) {
       throw new IllegalStateException(
           "Downstream dependencies analysis not activated, cannot inject annotations virtually!");
     }
-    try (BufferedOutputStream os =
-        new BufferedOutputStream(new FileOutputStream(libraryModelPath.toFile()))) {
-      Set<String> rows =
-          changes.stream()
-              .filter(addAnnotation -> addAnnotation.getLocation().isOnMethod())
-              .map(
-                  annot ->
-                      annot.getLocation().clazz
-                          + "\t"
-                          + annot.getLocation().toMethod().method
-                          + "\n")
-              .collect(Collectors.toSet());
+    // write methods
+    writeAnnotationsToFile(
+        changes.stream().filter(addAnnotation -> addAnnotation.getLocation().isOnMethod()),
+        libraryModelPath.resolve(NULLABLE_METHOD_LIST_FILE_NAME),
+        annot ->
+            Stream.of(
+                annot.getLocation().clazz + "\t" + annot.getLocation().toMethod().method + "\n"));
+    // write fields
+    writeAnnotationsToFile(
+        changes.stream().filter(addAnnotation -> addAnnotation.getLocation().isOnField()),
+        libraryModelPath.resolve(NULLABLE_FIELD_LIST_FILE_NAME),
+        annot ->
+            // An annotation on a single statement with multiple declaration will be considered for
+            // all declared variables. Hence, we have to mark all variables as nullable.
+            // E.g. for
+            // class Foo {
+            //     @Nullable Object a, b;
+            // }
+            // we have to consider both a and b be nullable and write each one
+            // ("Foo\ta" and "Foo\tb")
+            // on a separate line.
+            annot.getLocation().toField().variables.stream()
+                .map(variable -> annot.getLocation().clazz + "\t" + variable + "\n"));
+  }
+
+  /**
+   * Writes the passed annotation to the passed file. It uses the passed mapper to map the
+   * annotation to a string. And writes each string to a separate line.
+   *
+   * @param annotations Annotations to be written.
+   * @param path Path to the file to be written.
+   * @param mapper Mapper to map the annotation to a string.
+   */
+  private static void writeAnnotationsToFile(
+      Stream<AddAnnotation> annotations,
+      Path path,
+      Function<AddAnnotation, Stream<String>> mapper) {
+    try (BufferedOutputStream os = new BufferedOutputStream(new FileOutputStream(path.toFile()))) {
+      Set<String> rows = annotations.flatMap(mapper).collect(Collectors.toSet());
       for (String row : rows) {
         os.write(row.getBytes(Charset.defaultCharset()), 0, row.length());
       }
       os.flush();
```

Comment:
Not a big deal, but I don't think this call is needed, since the stream will immediately be closed afterward

Reply:
Sure, fixed in  f775354be1a90e45c7f67ffa5cdb0ecfe7010488

Reply:
I still see the call to `os.flush()`?  Why is this needed?

Reply:
Is the call to `os.flush()` redundant (or guaranteed to be executed after the block) in `try-with-resources` block ? IntelliJ only grays out call to `os.close()` but not for `os.flush()`.

Reply:
`try-with-resources` guarantees the stream will be closed, and I believe it will be flushed automatically before it is closed.

Reply:
Ok thanks for that. Then we can remove it. Fixed in 24f0a244677677c0cc7a15a154eda18bd8425080